### ISSUE TITLE
Firebase calls iOS9+ method on iOS 8

### ISSUE THREAD

**Issue by ghugues** [firebase/firebase-ios-sdk#1536]
### Environment

  * Xcode version: 9.2
  * Firebase SDK version: 5.1
  * Firebase Components: FirebaseDynamicLinks (3.0.1)

### Problem

On iOS 8, Firebase calls `-[UIApplicationDelegate application:openURL:options:]`. This method was introduced in iOS 9 and should not be called on iOS 8.

In particular, if this method has been marked as available only in iOS 9 or later, then the compiler will not warn about usage of iOS9+ APIs like `options[UIApplicationOpenURLOptionsSourceApplicationKey]`. Calling such code on iOS 8 where `UIApplicationOpenURLOptionsSourceApplicationKey` will resolve as `NULL` will crash the app.

#### Steps to reproduce:

 - Add Firebase DynamicLinks to an app that targets iOS 8 and later.
 - Setup a custom url scheme
 - Implement both `-[UIApplicationDelegate application:openURL:sourceApplication:annotation:]` and `-[UIApplicationDelegate application:openURL:options:]`
 - Launch this app on iOS 8

#### Stacktrace

```
Thread 1 Queue : com.apple.main-thread (serial)
  #0	0x0021a15a in -[XXXAppDelegate application:openURL:options:]
  #1	0x006abb51 in -[FIRAAppDelegateProxy application:openURL:options:] ()
  #2	0x00733c8b in -[FIRDynamicLinks passRetrievedDynamicLinkToApplication:] ()
  #3	0x00733e43 in -[FIRDynamicLinks retrievalProcess:completedWithResult:] ()
  #4	0x0072d739 in -[FIRDLDefaultRetrievalProcessV2 handleRequestResultsUpdated] ()
  #5	0x0072ce6f in __68-[FIRDLDefaultRetrievalProcessV2 retrievePendingDynamicLinkInternal]_block_invoke ()
  #6	0x00731a94 in __105-[FIRDynamicLinkNetworking sendRequestWithBaseURLString:requestBody:endpointPath:parserBlock:completion:]_block_invoke.183 ()
  #7	0x0749530a in _dispatch_call_block_and_release ()
  #8	0x074b5e2f in _dispatch_client_callout ()
  #9	0x0749c90e in _dispatch_main_queue_callback_4CF ()
  #10	0x0223995e in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
  #11	0x021f8760 in __CFRunLoopRun ()
  #12	0x021f7bcb in CFRunLoopRunSpecific ()
  #13	0x021f79fb in CFRunLoopRunInMode ()
  #14	0x0811b24f in GSEventRunModal ()
  #15	0x0811b08c in GSEventRun ()
  #16	0x040608b6 in UIApplicationMain ()
  #17	0x000dddcc in main
  #18	0x074e0ac9 in start ()
Enqueued from NSOperationQueue 0x7bffb5a0 :: NSOperation 0x7cc3f5d0 (Thread 17) Queue : NSOperationQueue 0x7bffb5a0 :: NSOperation 0x7cc3f5d0 (serial)
  #0	0x074b7603 in _dispatch_introspection_queue_item_enqueue ()
  #1	0x07498a05 in _dispatch_barrier_async_f ()
  #2	0x074991b4 in dispatch_async ()
  #3	0x00731994 in __105-[FIRDynamicLinkNetworking sendRequestWithBaseURLString:requestBody:endpointPath:parserBlock:completion:]_block_invoke ()
  #4	0x0072fbed in __FIRMakeHTTPRequest_block_invoke ()
  #5	0x01a65c03 in __49-[__NSCFLocalSessionTask _task_onqueue_didFinish]_block_invoke ()
  #6	0x02ec2715 in __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ ()
  #7	0x02de9165 in -[NSBlockOperation main] ()
  #8	0x02dc8567 in -[__NSOperationInternal _start:] ()
  #9	0x02dc8299 in -[NSOperation start] ()
  #10	0x02dc80e3 in __NSOQSchedule_f ()
  #11	0x074b5e2f in _dispatch_client_callout ()
  #12	0x0749bafc in _dispatch_queue_drain ()
  #13	0x0749b3c3 in _dispatch_queue_invoke ()
  #14	0x0749e067 in _dispatch_root_queue_drain ()
  #15	0x0749f84a in _dispatch_worker_thread3 ()
  #16	0x07800fa5 in _pthread_wqthread ()
  #17	0x07800a32 in start_wqthread ()
```

#### Going further

Although this is another subject, I would argue that Firebase should not even attempt to call `-[UIApplicationDelegate application:openURL:sourceApplication:annotation:]` or `-[UIApplicationDelegate application:openURL:options:]`, but should use `-[UIApplication openURL:]` or `-[UIApplication openURL:options:completionHandler:]` instead which are the official way of opening an URL (and opening the url scheme of the current application is perfectly legal and will behave as expected).

Also: It seems that Firebase is using this as a way to pass some information around (`deep-link-url-scheme://google/link/?dismiss=1&is_weak_match=1`). Maybe in an attempt to reach the `DynamicLinks` component ? Why does it need to do that ? Isn't there a better way for Firebase to pass around private information ?
I'm asking because this makes it much harder to track the app launch source (whether it was opened from a deep link), because we need to filter out these Firebase calls (which are undocumented, so hard to predict and catch exhaustively).

**Comment by paulb777** [comment]
At first glance, it looks like Firebase makes the assumption that if `-[UIApplicationDelegate application:openURL:options:]` exists in the app, the API exists and is valid to call. 

We'll follow up with a more detailed analysis and response.

**Comment by baolocdo** [comment]
We will investigate this. In the meantime, to avoid crashes, please set the flag `FirebaseAppDelegateProxyEnabled` to `NO` (Boolean) in the Info.plist.

**Comment by ghugues** [comment]
Thanks. We're already doing an iOS version check inside the method to avoid potential crashes. I just think this is something that could be improved in the Firebase SDK. And it's not just this method, I think that any method added to the protocol in a given iOS version should not be called on earlier versions, even if it's implemented (because it's going to be unexpected by the developer most of the time).
But you could easily argue that if the method is part of public API - which it is - and is implemented, then you can call it.

**Comment by baolocdo** [comment]
So I have just tested with iOS 8 with Firebase and Analytics SDKs that only implements iOS 9 openURL handler and I couldn't reproduce similar result. The iOS 9 was not called. I'm suspecting it's the Analytics problem, and Dynamic Links might behave differently. Can you provide the Podfile? Thanks.

**Comment by baolocdo** [comment]
@ghugues : Have you tried unlinking DynamicLinks? If it works as intended after taking DynamicLinks, I suspect it is the problem, not Analytics.

**Comment by ghugues** [comment]
I haven't, but yes the problem most probably comes from `DynamicLinks`. I initially framed this as an `Analytics` bug because the call was coming from `FIRAAppDelegateProxy`, but the stack trace clearly shows that the culprit is indeed `DynamicLinks`. I have updated the issue description since then.

**Comment by baolocdo** [comment]
Please update to the latest SDK versions and let us know whether it is still a problem. Thanks

**Comment by ryanwilson** [comment]
The fix was released as part of Firebase 5.5.0 - please re-open if you're still seeing the issue, and thanks for reporting it!
